src/v3/src/transformer/field/transform.ts

```diff
@@ -52,8 +52,27 @@ export const transformStepInputs: StepTransformer = (step?: NextStep): FormBag =
     }, [])
     .filter((input) => input.visible !== false && input.mutable !== false)
     .reduce((acc: FormBag, input: Input) => {
+      const { name, required, mutable } = input;
+
+      // add uischema
       const uischema = mapUiElement(input);
       acc.uischema.elements = [...acc.uischema.elements, uischema];
+
+      // add client validation for "required" field
+      // do not validate immutable fields, they will always be added to payload programatically
+      if (required && mutable !== false) {
+        acc.dataSchema[name] = {
+          validate(data) {
+            const isValid = !!data[name];
+            return isValid ? undefined : {
+              i18n: {
+                key: 'model.validation.field.blank',
+              },
+            };
+          },
+        };
+      }
```

Comment:
Instead of introducing `dataSchema`, can you build up the schema property that already exists? 

It would actually be extremely simple, since this is only handling `required` properties, you can build it up like so: 

```
if (required && mutable !== false) {
   acc.schema.properties[name] = {
      type: input.type
   }
   acc.schema.required.push(name);
}
```

Then, you can purely validate based on the required properties (using a simple function). But this will adhere to JsonSchema architecture and make it extremely easy to drop in AJV when we are ready.